explainer.md

```diff
@@ -182,565 +179,259 @@ Web apps that push the boundaries of what's possible on the web struggle to make
-
-However, one important consideration is that in current web accessibility specs,
-whether or not a value is legal or not often depends on the context.
-
-For example,
-the **`checked`** property is only defined on accessible nodes with certain roles
-like **`checkbox`** and **`radio`**.
-Therefore it's illegal to try to set **`checked`** on an object with a role where it isn't defined,
-like **`heading`**.
-However, it seems like it would be unfortunate
-if setting `checked=true` and then setting `role="checkbox"` would fail,
-while doing it in the other order would succeed.
-
-So instead, we propose the following rule:
-
-**When setting the value of an accessible property, that value is
-preserved in the accessible node's internal state. At the time the
-accessible property is accessed, illegal values are ignored.**
-
-First let's see an example of how that applies in practice when trying to set a role.
-
-```js
-var heading = document.createElement("h1");
-var axHeading = heading.accessibleNode;
-axHeading.role;  // returns "heading"
-
-axHeading.role = "button";
-axHeading.role;  // returns "button" because that's a legal role
+<!-- Page author uses the custom element as they would a native element -->
+<custom-slider min="0" max="5" value="3"></custom-slider>
 
-axHeading.role = "captain";
-axHeading.role;  // returns "heading" because "captain" is not legal.
+<!-- Custom element is forced to "sprout" extra attributes to express semantics -->
+<custom-slider min="0" max="5" value="3" role="slider"
+               tabindex="0" aria-valuemin="0" aria-valuemax="5"
+               aria-valuenow="3" aria-valuetext="3"></custom-slider>
 ```
 
-If ARIA attributes are used, the same behavior applies,
-in that accessing the **`role`** property returns the *computed* property value,
-which is taken:
-* from the accessible node first,
-* from the ARIA attribute second,
-* and from the HTML element's native semantics third.
+Using AOM, a Custom Element author would be able to write a class declaration like this:
 
 ```js
-var textbox = document.createElement("input");
-var axTextbox = textbox.accessibleNode;
-axTextbox.role;  // returns "textbox"
-
-textbox.setAttribute("role", "combobox");
-axTextbox.role;  // returns "combobox"
-
-textbox.setAttribute("role", "Portia");
-axTextbox.role;  // returns "textbox" because "Portia" is not legal.
-
-textbox.setAttribute("role", "combobox");
-axTextbox.role = "searchbox";
-axTextbox.role;  // returns "searchbox" because the AOM overrides ARIA.
+class CustomCheckbox extends HTMLElement {
+  constructor() {
+    super();
+
+    // Apply a role of "checkbox" via the AOM
+    this.accessibleNode.role = "checkbox";
+  }
+
+  // Observe the custom "checked" attribute
+  static get observedAttributes() { return ["checked"]; }
+
+  // When the custom "checked" attribute changes,
+  // keep the accessible checked state in sync.
+  attributeChangedCallback(name, old, newValue) {
```

Comment:
@marcoscaceres In case you have not figured it out yet, `attributeChangedCallback` is part of the [Custom Elements](https://w3c.github.io/webcomponents/spec/custom/#custom-element-reactions) API. According to the draft, there are a couple of callback functions an author can define upon the custom element's constructor's prototype object, to respond to certain occurrences related to an element. If an attribute of a custom element is changed, removed, etc., and the constructor provides a function named `attributeChangedCallback`, the function will be invoked, so that the custom element can adapt to the state change if necessary.